doc: quickstart for Helm projects and more details for the migration guide (#3552)
operator-framework/operator-sdk: website/content/en/docs/building-operators/helm/migration.md

$ operator-sdk init --plugins=helm --domain=example.com
```

> **Note** Ensure that you use the same values for the flags to recreate the same Helm chart and API. If you have
> more than one chart or API you can add them via `operator-sdk create api` command. For further information check the [Helm Quickstart][quickstart]. 

Reviewer: I think this change makes the doc less clear. I think we should revert this, and add an extra sentence about `--version` and `--kind`

```suggestion
Now that we have our new project initialized, we need to re-create each of our APIs.
Using our API example from earlier (`cache.example.com`), we'll use `cache` for the
`--group` flag.

For `--version` and `--kind`, we use `spec.versions[0].name` and `spec.names.kind`, respectively.
```

For each API in the existing project, run:

> ```sh

Reviewer: This block needs to be updated. There is no longer an `--api-version` flag.

### Checking the Permissions (RBAC)

In your new project, roles are automatically generated in `config/rbac/role.yaml`.
If you modified these permissions manually in `deploy/role.yaml` in your existing
project, you need to re-apply them in `config/rbac/role.yaml`.

New projects are configured to watch all namespaces by default, so they need a `ClusterRole` to have the necessary permissions. Ensure that `config/rbac/role.yaml` remains a `ClusterRole` if you want to retain the default behavior of the new project conventions. For further information refer to the [operator scope][operator-scope] documentation.  

The following rules were used in earlier versions of helm-operator to automatically create and manage services and servicemonitors for metrics collection. If your operator's charts don't require these rules, they can safely be left out of the new `config/rbac/role.yaml` file:

```  
  - apiGroups:
    - monitoring.coreos.com
    resources:
    - servicemonitors
    verbs:
    - get
    - create
  - apiGroups:
    - apps
    resourceNames:
    - memcached-operator
    resources:
    - deployments/finalizers
    verbs:
    - update
```

### Configuring your Operator

> In case your project has customizations in the `deploy/operator.yaml` then, it needs to be port to 
> `config/manager/manager.yaml`. However, note that the following env vars are no longer used. 

Reviewer: ```suggestion
If your existing project has customizations in `deploy/operator.yaml`, they need to be ported to 
`config/manager/manager.yaml`. If you are passing custom arguments in your deployment, make sure to also update `config/default/auth_proxy_patch.yaml`.

Note that the following environment variables are no longer used. 
```

 
### Migrating your CR's

Update the CR manifests in `config/samples` with the values of the CR's in your old project which are in `deploy/crds/`.

### Migrating your Watches
Check if you have customizations options in the `watch.yaml` file of your previous project and then, update the new `watch.yaml` file with the same ones. In our example, it will look like:

```yaml
# Use the 'create api' subcommand to add watches to this file.
- group: example.com
  version: v1alpha1
  kind: Nginx
  chart: helm-charts/nginx
# +kubebuilder:scaffold:watch
```

**NOTE**: Do not remove the [Maker][marker] `+kubebuilder:scaffold:watch`. It allows the tool to update the watches file when new API's are created. 

### Checking the Permissions (RBAC)

SDK generated the roles for you in the `config/rbac/role.yaml`. However, you might need to  
update this file with the permissions which are in your previous project in `/deploy/roles.yaml`.

And then, beware that:

- By default, the projects are `cluster-scoped`. Check that your role is of the `ClusterRole` Kind. For further information check the [operator-scope][operator-scope] doc.  

> - The `apiGroups` for `monitoring.coreos.com` is no longer required. If you would like to export the Prometheus metrics check the next section `Exporting metrics`.

Author: All perfect :-) But I still think that we need to add a note to highlight that `The `apiGroups` for `monitoring.coreos.com` is no longer required` otherwise, could not users face permissions issues and decide to port it thinking that it is the root cause?

> - The `apiGroups` for `monitoring.coreos.com` is no longer required. If you would like to export the Prometheus metrics check the next section `Exporting metrics`.

Author: I just made a change here. Instead o describe the actions/resource I added the piece from the yaml file. I think it will make the user's life easier.

### Migrating your Helm Charts

> If you have customizations in your `helm-charts` then, apply them in the new `helm-charts`. Note that this directory was not changed at all.

Author: But it will create the api based in the chart. However, imagine the scenario:
- The dev creates the project and the api based on chart
- Then start to work on it and then, did customizations directly in the `helm-charts` dir of the project and no longer in the specific chart. Could not that be common?

> If you have customizations in your `helm-charts` then, apply them in the new `helm-charts`. Note that this directory was not changed at all.

Reviewer: It would be common, but any of those customizations would be copied over automatically if they use `create api --helm-chart=<existing_project>/helm-charts/<chart>`

> If you have customizations in your `helm-charts` then, apply them in the new `helm-charts`. Note that this directory was not changed at all.

Reviewer: > The scenario that I am trying to cover here is when: the changes were applied directly in the operator project and not in the chart used to gen the API.

Yes that's the scenario that's covered by copying the helm chart from the existing project.

### Configuring your Operator

In case your project has customizations in the `deploy/operator.yaml` then, it needs to be port to 
`config/manager/manager.yaml`. However, note that the following env vars are no longer supported. 

- `OPERATOR_NAME`: Now, the name of the operator can be defined due the flag `--project-name` when the project is 
created. By default, it will be the current directory name. Also, see that it is stored in the PROJECT file. 
- `WATCH_NAMESPACE`: By ddefault, it will be configured as `cluster-scoped`. To change it, use the flag `-leader-election-namespace` when you start the operator. 
- `POD_NAME`:  By default, it will be the projectName stored in the PROJECT file. To change it, use the flag `-leader-election-id` when you start the operator. 

> **Note** The Dockerfile image also changes and now it is `distroless` and still been `rootless`. You might need to port some customizations made in your old project Dockerfile as well.

Reviewer: This is incorrect. The Helm operator base image [still uses `ubi8/ubi-minimal`](https://github.com/operator-framework/operator-sdk/blob/35c871fa0ddf0ac473b20f58ac03e8897d6e38cd/hack/image/helm/Dockerfile#L1)

```suggestion
```

Author: Oh it is true.

### Checking the Permissions (RBAC)

In your new project, roles are automatically generated in `config/rbac/role.yaml`.
If you modified these permissions manually in `deploy/role.yaml` in your existing
project, you need to re-apply them in `config/rbac/role.yaml`.

New projects are configured to watch all namespaces by default, so they need a `ClusterRole` to have the necessary permissions. Ensure that `config/rbac/role.yaml` remains a `ClusterRole` if you want to retain the default behavior of the new project conventions. For further information refer to the [operator scope][operator-scope] documentation.  

The following rules were used in earlier versions of helm-operator to automatically create and manage services and servicemonitors for metrics collection. If your operator's charts don't require these rules, they can safely be left out of the new `config/rbac/role.yaml` file:

```  
  - apiGroups:
    - monitoring.coreos.com
    resources:
    - servicemonitors
    verbs:
    - get
    - create
  - apiGroups:
    - apps
    resourceNames:
    - memcached-operator
    resources:
    - deployments/finalizers
    verbs:
    - update
```

### Migrating your Helm Charts

> Above, you already created the API with the Helm-Chart using the command `operator-sdk create api`. However, if you have in your previous project customizations which were applied just in the operator and NOT in the `Helm-chart` used to create the API then, ensure that you will migrate them to the new project as well. 

Reviewer: I still disagree with this section. If I run `operator-sdk create api --helm-chart=<existing_project>/helm-charts/<chart>`, as is documented earlier in this doc, there will be no further helm chart migrations necessary.

The `./helm-charts` dirs in both old and new projects will be identical.

Reviewer: Using a directory path with the `--helm-chart` flag effectively just copies the directory into the new project.

Author: Yep. It is true.
